gap/algobj.gi

```diff
@@ -0,0 +1,18 @@
+# Global bindings used for graph colouring algorithm selection.
+DigraphColouringAlgorithmFamilyObj := NewFamily(
+                                      "DigraphColouringAlgorithmFamily",
+                                      IsDigraphAlgorithm,
+                                      IsDigraphColouringAlgorithm);
+BindGlobal("DigraphColouringAlgorithmLawler",
+            Objectify(
+              NewType(
+                DigraphColouringAlgorithmFamilyObj,
+                IsDigraphColouringAlgorithmLawler),
+              rec()));
+BindGlobal("DigraphColouringAlgorithmByskov",
+            Objectify(
+              NewType(
+                DigraphColouringAlgorithmFamilyObj,
+                IsDigraphColouringAlgorithmByskov),
+              rec()));
+Unbind(DigraphColouringAlgorithmFamilyObj);
```

Comment:
Why have you chosen to go for this setup? This way, `DigraphColouringAlgorithmLawler` (for instance) is just an object that basically only knows its name, and that it's a name representing an algorithm. Were there some discussions on this topic this that I missed out on?

I've not thought this through... and I'm not sure this is even possible... but... another option could be that `IsDigraphAlgorithm` objects (amongst other things) are actual functions (or at least they contain a function) that implement the algorithm in question. So `DigraphColouringAlgorithmLawler` could be an object satisfying `IsDigraphAlgorithm`, but at the same time it contains its own implementation, so you could do `DigraphColouringAlgorithmLawler(D)` for a digraph `D` if you really wanted and get the result of applying the algorithm to the graph.

And then you'd just have the one method:
```
InstallMethod(ChromaticNumber, "for a digraph and colouring algorithm",
[IsDigraph, IsDigraphColouringAlgorithm],
{D, f} -> f(D));
```
What do you think of that? I'm just curious if you had thought of that, or if you or anyone else (@mtorpey?) have opinions on the best setup.

Reply:
We discussed what the best setup would be with @james-d-mitchell a few months ago, and we settled on a setup which is what Ewan's implemented here: a filter for each algorithm, which basically just knows its own name, to allow a user to specify the right one.  However, I don't think any of us considered actually having the algorithm **be** the function like @wilfwilson suggests.

This suggestion seems to make a lot of sense!  It'd be less surprising, more "object-oriented" (as far as anything in GAP is!) and would still let us set up filters in interesting ways later on.

@EwanGilligan and @james-d-mitchell, does Wilf's suggestion sound good to you?

Reply:
Thanks for that context @mtorpey! (And just to be clear: my suggestion isn't a request - I don't really mind what the setup is.)

Reply:
Like @mtorpey has said, I don't think we thought about the suggestion of the object passed in be the function was considered. To me this would make more sense, but I'll wait to hear what @james-d-mitchell has to say before I make the change.

Reply:
@wilfwilson This has been discussed a bit further now, with the outcome mostly being that passing a function would be nicer, but the current method works better due to the way GAP filters are set up.

So the main issue with the approach of passing the function directly to `ChromaticNumber` is that you would not be able to use filters to restrict the function that could be passed. As functions are internal representations, it is not possible to set filters on them after creation. It would be possible to wrap this function in an object that contains the function and has the right filters, but then you would lose the ability to call the function directly. This would be a nice solution if GAP allowed you to override the call method, but unfortunately this is not possible.

So then the option to store an object that only knows it's own name or storing an object that stores the function would be identically used by the user. The current mechanism is then a bit simpler, as the function wrapping way would require an additional method to be installed that just called the wrapped function. I've found out that the name of the current mechanism is something known as tagged dispatch, so is a common pattern for these types of problems.

Hope this ramble has helped clear up why we've chose to perform the method selection this way.

Reply:
That's fine @EwanGilligan 🙂. Nevertheless, this might be interesting...
> This would be a nice solution if GAP allowed you to override the call method, but unfortunately this is not possible.

I learned this morning that (I think?!)  you _can_ override the call method - i.e. look at this:
```
gap> InstallMethod(CallFuncList, [IsDigraph, IsList], {f, args} -> Sum(args));
gap> D := RandomDigraph(10);;
gap> D(4, 6, 2);
12
```
So you can 'call' a digraph on some arguments! (Obviously nonsense.) But if your wrapper objects were called `IsEwanWrapper`, and they were component objects with the component `func` being the underlying function, then you could do:
```
gap> InstallMethod(CallFuncList, [IsEwanWrapper, IsList], {f, args} -> CallFuncList(f!.func, args));
```
So if `f` was one of these wrapper objects, you could do `f(x)` and this call would be passed to `f!.func`, i.e. the result of `f(x)` would be `f!.func(x)`.

Reply:
But again, I'm not requesting that you do this 🙂 Thanks for thinking about this and replying thoughtfully. I'm happy with whatever.

Reply:
@wilfwilson Very interesting about overriding `CallFuncList`, I'd asked around but nobody seemed to know that it was possible.

I think overall the current "tagged dispatch" approach is probably simpler, as it doesn't require the additional method to override CallFuncList for the wrapper object. Thank you for taking the time to let me know about the other option